plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiPageSource.java

```diff
@@ -13,36 +13,87 @@
 import io.trino.plugin.hive.HiveColumnHandle;
 import io.trino.spi.Page;
 import io.trino.spi.block.Block;
 import io.trino.spi.block.RunLengthEncodedBlock;
 import io.trino.spi.connector.ConnectorPageSource;
+import org.apache.hadoop.fs.Path;
 
 import java.io.IOException;
 import java.util.List;
 import java.util.Map;
 
+import static io.airlift.slice.Slices.utf8Slice;
 import static io.trino.plugin.base.util.Closables.closeAllSuppress;
+import static io.trino.plugin.hive.HiveColumnHandle.FILE_MODIFIED_TIME_COLUMN_NAME;
+import static io.trino.plugin.hive.HiveColumnHandle.FILE_MODIFIED_TIME_TYPE_SIGNATURE;
+import static io.trino.plugin.hive.HiveColumnHandle.FILE_SIZE_COLUMN_NAME;
+import static io.trino.plugin.hive.HiveColumnHandle.FILE_SIZE_TYPE_SIGNATURE;
+import static io.trino.plugin.hive.HiveColumnHandle.PARTITION_COLUMN_NAME;
+import static io.trino.plugin.hive.HiveColumnHandle.PARTITION_TYPE_SIGNATURE;
+import static io.trino.plugin.hive.HiveColumnHandle.PATH_COLUMN_NAME;
+import static io.trino.plugin.hive.HiveColumnHandle.PATH_TYPE;
+import static io.trino.spi.predicate.Utils.nativeValueToBlock;
+import static io.trino.spi.type.DateTimeEncoding.packDateTimeWithZone;
+import static io.trino.spi.type.TimeZoneKey.UTC_KEY;
 import static java.util.Objects.requireNonNull;
 
 public class HudiPageSource
         implements ConnectorPageSource
 {
-    private final List<HiveColumnHandle> columnHandles;
+    private final Block[] prefilledBlocks;
+    private final int[] delegateIndexes;
     private final ConnectorPageSource dataPageSource;
-    private final Map<String, Block> partitionBlocks;
 
     public HudiPageSource(
+            String partitionName,
             List<HiveColumnHandle> columnHandles,
             Map<String, Block> partitionBlocks,
-            ConnectorPageSource dataPageSource)
+            ConnectorPageSource dataPageSource,
+            Path path,
+            long fileSize,
+            long fileModifiedTime)
     {
-        this.columnHandles = ImmutableList.copyOf(requireNonNull(columnHandles, "columnHandles is null"));
-        this.partitionBlocks = ImmutableMap.copyOf(requireNonNull(partitionBlocks, "partitionBlocks is null"));
+        requireNonNull(columnHandles, "columnHandles is null");
         this.dataPageSource = requireNonNull(dataPageSource, "dataPageSource is null");
+
+        int size = columnHandles.size();
+        this.prefilledBlocks = new Block[size];
+        this.delegateIndexes = new int[size];
+
+        int outputIndex = 0;
+        int delegateIndex = 0;
+
+        for (HiveColumnHandle column : columnHandles) {
+            if (partitionBlocks.containsKey(column.getName())) {
+                Block partitionValue = partitionBlocks.get(column.getName());
+                prefilledBlocks[outputIndex] = partitionValue;
+                delegateIndexes[outputIndex] = -1;
+            }
+            else if (column.getName().equals(PARTITION_COLUMN_NAME)) {
```

Comment:
I also notice that there are few other internal columns in `HiveColumnHandle` such as `$bucket` and `$row_id`. When do we need to those columns?

Reply:
The `$bucket` simply provides the bucket number of the current row. `$row_id` is an internal column for `UPDATE` and `MERGE` statements. The former is for users, latter is for internal usage.